File: day_2/4/solution.py
import numpy as np

def main():
    data = np.loadtxt("input", dtype=str)
    directions = data[:, 0]
    movement = data[:, 1].astype(int)
    horizontal_position = 0
    depth = 0
    aim = 0
    for direction, movement in zip(directions, movement):
        if direction == "up":
            aim -= movement
        elif direction == "down":
            aim += movement
        elif direction == "forward":
            horizontal_position += movement
            depth += aim * movement
    print(depth, horizontal_position, depth * horizontal_position)

if __name__ == "__main__":
    main()

File: day_2/4/solution.go
package main

import (
	"fmt"
	"os"
	"log"
	"bufio"
	"strings"
	"strconv"
)

func main() {
	file, err := os.Open("input")
	if err != nil {
		log.Fatal(err)
	}
	scanner := bufio.NewScanner(file)
	horizontal_position := 0
	aim := 0
	depth := 0
	for scanner.Scan() {
		text := scanner.Text()
		line := strings.Split(text, " ")
		direction := line[0]
		_movement, err := strconv.ParseInt(line[1], 10, 0)
		if err != nil {
			log.Fatal(err)
		}
		movement := int(_movement)
		if direction == "forward" {
			horizontal_position += movement
			depth += movement * aim
		} else if direction == "up" {
			aim -= movement
		} else if direction == "down" {
			aim += movement
		} else {
			log.Fatal("bad direction")
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Printf("%d %d %d\n", depth, horizontal_position, horizontal_position * depth)
}
